src/blueapi/service/interface.py

```diff
@@ -20,52 +18,81 @@
 context and worker"""
 
 
-class _Singleton:
-    context: BlueskyContext
-    worker: Worker
-    messaging_template: MessagingTemplate | None = None
-    initialized = False
+@runtime_checkable
+class CacheClearable(Protocol):
+    def cache_clear(self) -> None: ...
 
 
-def start_worker(
-    config: ApplicationConfig,
-    bluesky_context: BlueskyContext = None,
-    worker: TaskWorker = None,
-) -> None:
-    """Creates and starts a worker with supplied config"""
-    if _Singleton.initialized:
-        raise Exception("Worker is already running. To reload call stop first")
-    _Singleton.context = bluesky_context
-    if _Singleton.context is None:
-        _Singleton.context = BlueskyContext()
-        _Singleton.context.with_config(config.env)
-    _Singleton.worker = worker
-    if _Singleton.worker is None:
-        _Singleton.worker = TaskWorker(
-            _Singleton.context,
-            broadcast_statuses=config.env.events.broadcast_status_events,
-        )
-    if config.stomp is not None:
-        _Singleton.messaging_template = StompMessagingTemplate.autoconfigured(
-            config.stomp
-        )
+_CONFIG: ApplicationConfig = ApplicationConfig()
 
-    # Start worker and setup events
-    _Singleton.worker.start()
-    if _Singleton.messaging_template is not None:
-        event_topic = _Singleton.messaging_template.destinations.topic(
-            "public.worker.event"
-        )
+
+def config() -> ApplicationConfig:
+    return _CONFIG
+
+
+def set_config(new_config: ApplicationConfig):
+    global _CONFIG
+
+    _CONFIG = new_config
+
+
+@lru_cache
+def context() -> BlueskyContext:
+    ctx = BlueskyContext()
+    ctx.with_config(config().env)
+    return ctx
+
+
+@lru_cache
+def worker() -> Worker:
+    worker = TaskWorker(
+        context(),
+        broadcast_statuses=config().env.events.broadcast_status_events,
+    )
+    worker.start()
+    return worker
+
+
+@lru_cache
+def messaging_template() -> MessagingTemplate | None:
+    stomp_config = config().stomp
+    if stomp_config is not None:
+        template = StompMessagingTemplate.autoconfigured(stomp_config)
+
+        task_worker = worker()
+        event_topic = template.destinations.topic("public.worker.event")
 
         _publish_event_streams(
             {
-                _Singleton.worker.worker_events: event_topic,
-                _Singleton.worker.progress_events: event_topic,
-                _Singleton.worker.data_events: event_topic,
+                task_worker.worker_events: event_topic,
+                task_worker.progress_events: event_topic,
+                task_worker.data_events: event_topic,
             }
         )
-        _Singleton.messaging_template.connect()
-    _Singleton.initialized = True
+        template.connect()
+        return template
+    else:
+        return None
+
+
+def setup(config: ApplicationConfig) -> None:
```

Comment:
Do we want to prevent this being called twice (without teardown)?

I think if it was called a second time with a different config, the config would be set but the worker would still be cached and using the old config?